### ISSUE TITLE
rewrite mutable<->readonly assignment rules in terms of is-immutable-type lib?

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#708]
Being able to throw away the reams and reams of dodgy AST wrangling that I've written and replacing with https://github.com/RebeccaStevens/is-immutable-type/ (which looks great) would be a nice outcome

**Comment by danielnixon** [comment]
Example (new) rule `no-decreasing-immutability`: https://github.com/danielnixon/eslint-plugin-total-functions/pull/new/use-is-immutable-type

This only checks assignment expressions, and it doesn't have any config options. But it demonstrates how we might throw away the accumulated cruft and replace with `is-immutable-type`.

**Comment by danielnixon** [comment]
A downside to this approach is that we would lose fine-grained property-wise (and parameter-wise) comparisons.

This wouldn't matter for assignments that change the immutability level (Immutable -> DeepReadonly or DeepReadonly -> ShallowReadonly or anything to Mutable), but it would allow an assignment like the following, which should arguably be prevented **even though both source and destination types are the same immutability level**:

```typescript
type Foo = {
  readonly a: string;
  b: string;
};

type Bar = {
  a: string;
  readonly b: string;
};

const foo: Foo = { a: "a", b: "b" };

// Permitted because left and right hand side of assignment have same immutability level (both mutable)
const bar: Bar = foo;

bar.a = "whoops";

// Logs "whoops"
console.log(bar.a);
```

The only way to catch that would be to preserve our own AST wrangling that compares types in a property-wise (and parameter-wise) way. But it sure is tempting to be able to throw it all away and benefit from the momentum behind eslint-plugin-functional (including bug fixes, attention to linting performance, etc).

And how likely is it that someone is assigning from and to mutable types that are "mismatched" like this (contrived?) example? Probably low.

**Comment by danielnixon** [comment]
Individual parts of unions and intersections would probably be lost this way too. So much like the above example, you could (I suspect) assign `MutableA | ImmutableB` to `ImmutableA | MutableB`, which again should arguably be forbidden.

**Comment by danielnixon** [comment]
`is-immutable-type` doesn't look at call signatures, so if we ported to it we would lose the ability to flag sneaky mutation like this example:

```ts
type Data = {
  readonly value: string;
};

type DataHolder = {
  readonly a: Data;
  readonly doSomethingWithMyData: <T>(cb: (a: Data) => T) => T;
};

declare const foo: DataHolder;

foo.doSomethingWithMyData((a: { value: string }) => a.value = "whoops");
```
